metadata-ingestion/src/datahub/ingestion/source/dbt.py

```diff
@@ -0,0 +1,347 @@
+
+            catalog_nodes = dbt_catalog_json["nodes"]
+            catalog_sources = dbt_catalog_json["sources"]
+
+            all_catalog_entities = catalog_nodes | catalog_sources
+
+            logger.info(
+                "loading {} nodes, {} sources".format(
+                    len(all_manifest_entities), len(all_catalog_entities)
+                )
+            )
+
+            nodes = extract_dbt_entities(
+                all_manifest_entities, all_catalog_entities, platform, environment
+            )
+
+            return nodes
+
+
+def get_urn_from_dbtNode(
+    database: str, schema: str, name: str, platform: str, env: str
+) -> str:
+    db_fqn = f"{database}.{schema}.{name}".replace('"', "")
+    return f"urn:li:dataset:(urn:li:dataPlatform:{platform},{db_fqn},{env})"
+
+
+def get_custom_properties(node: DBTNode) -> Dict[str, str]:
+    tags = {}
+    tags["dbt_node_type"] = node.node_type
+    tags["materialization"] = node.materialization
+    tags["dbt_file_path"] = node.dbt_file_path
+    return tags
+
+
+def get_upstreams(
+    upstreams: List[str], all_nodes, platform: str, environment: str
+) -> List[str]:
+    upstream_urns = []
+
+    for upstream in upstreams:
+        upstream_node = all_nodes[upstream]
+
+        upstream_urns.append(
+            get_urn_from_dbtNode(
+                upstream_node["database"],
+                upstream_node["schema"],
+                upstream_node["name"],
+                platform,
+                environment,
+            )
+        )
+
+    return upstream_urns
+
+
+def get_upstream_lineage(upstream_urns: List[str]) -> UpstreamLineage:
+    ucl: List[UpstreamClass] = []
+
+    actor, sys_time = "urn:li:corpuser:dbt_executor", int(time.time()) * 1000
+
+    for dep in upstream_urns:
+        uc = UpstreamClass(
+            dataset=dep,
+            auditStamp=AuditStamp(actor=actor, time=sys_time),
+            type="TRANSFORMED",
+        )
+        ucl.append(uc)
+
+    ulc = UpstreamLineageClass(upstreams=ucl)
+
+    return ulc
+
+
+# This is from a fairly narrow data source that is posgres specific, we would expect this to expand over
+# time or be replaced with a more thorough mechanism
+_field_type_mapping = {
+    "boolean": BooleanTypeClass,
+    "date": DateTypeClass,
+    "numeric": NumberTypeClass,
+    "text": StringTypeClass,
+    "timestamp with time zone": DateTypeClass,
+    "integer": NumberTypeClass,
+}
+
+
+def get_column_type(
+    report: DBTSourceReport, dataset_name: str, column_type: str
+) -> SchemaFieldDataType:
+    """
+    Maps known DBT types to datahub types
+    """
+
+    pattern = re.compile(r"[\w ]+")  # drop all non alphanumerics
+    match = pattern.match(column_type)
+    if match is not None:
+        column_type_stripped = match.group()
+
+    TypeClass: Any = None
+    for key in _field_type_mapping.keys():
+        if key == column_type_stripped:
+            TypeClass = _field_type_mapping[column_type_stripped]
```

Comment:
I found this code a bit confusing. `column_type_stripped` is only defined if match is not none, which means the reference is undefined and will error if match is ever none.

Reply:
yah, that was the result of the mypy check noticing that match could be None so I added a test, I'm not familiar with mypy so it threw me for a bit of a loop.

I added an initializer which is also a bit confusing but at least it won't error out now.